serialization.rst

```diff
@@ -0,0 +1,337 @@
+  sent as an associative array to the ``bsonUnserialize`` function to initialise
+  the object's properties.
+
+  Remarks:
+
+  - If the ``__pclass`` property is not present [1]_, this throws the
+    ``MongoDB\Driver\Exception\UnexpectedValueException`` exception with a
+    message indicating the special class marker could not be found.
+
+  - If the class as determined by the ``__pclass`` property does not implement
+    the ``MongoDB\BSON\Persistable`` interface, throw an
+    ``MongoDB\Driver\Exception\UnexpectedValueException`` exception with a
+    message indicating that the named class does not implement the required
+    interface.
+
+- ``any other string`` — defines the class name that the BSON array or BSON
+  object should be deserialized at.
+
+  If the class implements the ``MongoDB\BSON\Unserializable`` interface, then
+  the properties of the BSON document, **minus** the ``__pclass`` property [1]_
+  if it exists, are sent as an associative array to the ``bsonUnserialize``
+  function to initialise the object's properties.
+
+  To discuss:
+
+  - What should we do if the named class implements
+    ``MongoDB\BSON\Persistable``?
+    
+    We can throw an ``MongoDB\Driver\Exception\UnexpectedValueException``
+    exception indicating that the BSON document really knows what PHP type it
+    should be. This could be important if the object needs to be deserialized
+    in a special way.
+
+    We can ignore it and act as the ``__pclass`` property wasn't set at all.
+    And not send the ``__pclass`` property as an array element to
+    ``bsonUnserialize``.
+
+TypeMaps
+--------
+
+TypeMaps can be set through the ``setTypeMap()`` on a
+``MongoDB\Driver\Cursor`` object, or the ``$typeMap`` argument of
+``MongoDB\BSON\toPHP()`` (previously, ``MongoDB\BSON\toArray()``). Each of the
+three classes (``root``, ``document`` and ``array``) can be individually set.
+If ``document`` is set to something else than the default and ``root`` is set
+to the default (or not set at all), then the value of ``document`` also counts
+for ``root``. Which makes the following example equivalent::
+
+    ->setTypeMap( [ 'document' => 'MyClass' ] )
+    ->setTypeMap( [ 'document' => 'MyClass', 'root' => 'MyClass' ] );
+    ->setTypeMap( [ 'document' => 'MyClass', 'root' => NULL ] );
+
+But these are **not** equivalent::
+
+    ->setTypeMap( [ 'root' => 'MyClass' ] )
+    ->setTypeMap( [ 'document' => 'MyClass', 'root' => 'MyClass' ] );
+
+In the first line of this second example, the value for ``document`` stays
+the default, which is ``stdClass``.
+
+Examples
+--------
+
+In these examples, ``MyClass`` does **not** implement any interface,
+``YourClass`` implements ``MongoDB\BSON\Unserializable`` and ``OurClass``
+implements ``MongoDB\BSON\Persistable``. Their ``bsonUnserialize()`` methods
+iterate over the array and set the properties without modifications. It
+**also** sets the ``unserialized = true`` property.
+
```

Comment:
Implementation in the YourClass and OurClass bsonUnserialize functions. It's to show in the example that that method was run. It's not some magic.